### PR TITLE
Courses are either optional or must be offered

### PR DESCRIPTION
**Pull request by czhang83**
Removed the constraint that all courses must be offered.
Added a column in the Course tab that says whether a course must be offered.
Changed the related error checking and print format.

### CODE REVIEW

File: schedule.py [hunk 1]
```diff
@@ -182,16 +184,18 @@ def read_input(input_file):
...
+    # check if the required units are more than professors' total units
     total_professor_units = sum(professor.max_units for professor in professors.values())
-    total_course_units = sum(section.units for section in sections.values())
-    if total_course_units > total_professor_units:
-        raise ValueError('Professors can only teach {} units but there are {} course units total'.format(
-            total_professor_units,
-            total_course_units,
-        ))
+    course_units_required = sum(section.units for section in sections.values() if section.must_offer)
+    course_units_optional = sum(section.units for section in sections.values() if not section.must_offer)
```
Comment (justinnhli) [inline schedule.py:190]: Okay, but I think the representation needs to change - I can imagine a semester requiring 2 sections of a course (which we do with CS1 and Data Structures), but can go up to 3 sections. So `must_offer` should be an integer, not a boolean.

File: schedule.py [hunk 2]
```diff
@@ -282,9 +277,15 @@ def print_results(solver, classes, professors, sections, semesters):
...
+                        if section.must_offer:
+                            print(section.name + ' assigned to ' + professor.name + ' (required) (requested)')
+                        else:
+                            print(section.name + ' assigned to ' + professor.name + ' (optional) (requested)')
                     else:
-                        print(section.name + ' assigned to ' + professor.name + ' (not requested)')
+                        if section.must_offer:
+                            print(section.name + ' assigned to ' + professor.name + ' (required) (not requested)')
+                        else:
+                            print(section.name + ' assigned to ' + professor.name + ' (optional) (not requested)')
```
Comment (justinnhli) [inline schedule.py:288]: I'm okay with leaving out "optional"/"required" here, which gets messy when there you require multiple sections.

File: schedule.py [hunk 3]
```diff
@@ -297,9 +298,15 @@ def print_results(solver, classes, professors, sections, semesters):
...
+                        if section.must_offer:
+                            print(professor.name + ' will be teaching ' + section.name + ' (required) (requested)')
+                        else:
+                            print(professor.name + ' will be teaching ' + section.name + ' (optional) (requested)')
                     else:
-                        print(professor.name + ' will be teaching ' + section.name + ' (not requested)')
+                        if section.must_offer:
+                            print(professor.name + ' will be teaching ' + section.name + ' (required) (not requested)')
+                        else:
+                            print(professor.name + ' will be teaching ' + section.name + ' (optional) (not requested)')
```
Comment (justinnhli) [inline schedule.py:309]: Ditto re: optional/required.

**Comment by czhang83** [comment]
@justinnhli Updated

**Pull request merged**